### ISSUE TITLE
mapcar: Symbol’s function definition is void: rest

### ISSUE THREAD

**Issue by r-darwish** [emacs-lsp/lsp-mode#665]
**Describe the bug**
LSP doesn't start. `*Messages*` shows:

```
error in process filter: mapcar: Symbol’s function definition is void: rest
error in process filter: Symbol’s function definition is void: rest
```

**To Reproduce**
Start emacs and open a buffer

**Expected behavior**
LSP should work

**Which Language Server did you use**
Python and Rust

**OS**
Both Windows and Arch Linux

**Error callstack**
```
Debugger entered--Lisp error: (void-function rest)
  rest((#s(lsp--workspace :parser #s(lsp--parser :headers nil :body nil :reading-body nil :body-length nil :body-received nil :leftovers "" :workspace #2) :server-capabilities nil :registered-server-capabilities nil :root "c:/Users/roey/src/topgrade/" :client #s(lsp--client :language-id nil :add-on? nil :new-connection (:connect #f(compiled-function (filter sentinel name) #<bytecode 0x294fcbd>) :test\? #f(compiled-function () #<bytecode 0x2951c51>)) :ignore-regexps nil :ignore-messages nil :notification-handlers #<hash-table equal 1/65 0x28a76dd> :request-handlers #<hash-table equal 0/65 0x2954cd1> :response-handlers #<hash-table eql 1/65 0x29ccf01> :prefix-function nil :uri-handlers #<hash-table equal 0/65 0x29ce351> :action-handlers #<hash-table equal 0/65 0x29cf01d> :use-native-json nil :major-modes (rust-mode rustic-mode) :activation-fn nil :priority -1 :server-id rls :multi-root nil :initialization-options nil :library-folders-fn nil :before-file-open-fn nil :initialized-fn nil :remote? nil :message-trace nil) :host-root nil :proc #<process rls> :cmd-proc #<process rls> :buffers (#<buffer generic.rs>) :highlight-overlays #<hash-table eq 0/65 0x2693b1d> :extra-client-capabilities nil :status starting :metadata #<hash-table equal 0/65 0x2693c45> :watches #<hash-table equal 0/65 0x2693c65> :workspace-folders nil :last-id 0 :status-string nil :shutdown-action nil :diagnostics #<hash-table equal 0/65 0x2693c85>) . #<hash-table equal 1/65 0x2633719>))
  mapcar(rest ((#s(lsp--workspace :parser #s(lsp--parser :headers nil :body nil :reading-body nil :body-length nil :body-received nil :leftovers "" :workspace #3) :server-capabilities nil :registered-server-capabilities nil :root "c:/Users/roey/src/topgrade/" :client #s(lsp--client :language-id nil :add-on? nil :new-connection (:connect #f(compiled-function (filter sentinel name) #<bytecode 0x294fcbd>) :test\? #f(compiled-function () #<bytecode 0x2951c51>)) :ignore-regexps nil :ignore-messages nil :notification-handlers #<hash-table equal 1/65 0x28a76dd> :request-handlers #<hash-table equal 0/65 0x2954cd1> :response-handlers #<hash-table eql 1/65 0x29ccf01> :prefix-function nil :uri-handlers #<hash-table equal 0/65 0x29ce351> :action-handlers #<hash-table equal 0/65 0x29cf01d> :use-native-json nil :major-modes (rust-mode rustic-mode) :activation-fn nil :priority -1 :server-id rls :multi-root nil :initialization-options nil :library-folders-fn nil :before-file-open-fn nil :initialized-fn nil :remote? nil :message-trace nil) :host-root nil :proc #<process rls> :cmd-proc #<process rls> :buffers (#<buffer generic.rs>) :highlight-overlays #<hash-table eq 0/65 0x2693b1d> :extra-client-capabilities nil :status starting :metadata #<hash-table equal 0/65 0x2693c45> :watches #<hash-table equal 0/65 0x2693c65> :workspace-folders nil :last-id 0 :status-string nil :shutdown-action nil :diagnostics #<hash-table equal 0/65 0x2693c85>) . #<hash-table equal 1/65 0x2633719>)))
  -map(rest ((#s(lsp--workspace :parser #s(lsp--parser :headers nil :body nil :reading-body nil :body-length nil :body-received nil :leftovers "" :workspace #3) :server-capabilities nil :registered-server-capabilities nil :root "c:/Users/roey/src/topgrade/" :client #s(lsp--client :language-id nil :add-on? nil :new-connection (:connect #f(compiled-function (filter sentinel name) #<bytecode 0x294fcbd>) :test\? #f(compiled-function () #<bytecode 0x2951c51>)) :ignore-regexps nil :ignore-messages nil :notification-handlers #<hash-table equal 1/65 0x28a76dd> :request-handlers #<hash-table equal 0/65 0x2954cd1> :response-handlers #<hash-table eql 1/65 0x29ccf01> :prefix-function nil :uri-handlers #<hash-table equal 0/65 0x29ce351> :action-handlers #<hash-table equal 0/65 0x29cf01d> :use-native-json nil :major-modes (rust-mode rustic-mode) :activation-fn nil :priority -1 :server-id rls :multi-root nil :initialization-options nil :library-folders-fn nil :before-file-open-fn nil :initialized-fn nil :remote? nil :message-trace nil) :host-root nil :proc #<process rls> :cmd-proc #<process rls> :buffers (#<buffer generic.rs>) :highlight-overlays #<hash-table eq 0/65 0x2693b1d> :extra-client-capabilities nil :status starting :metadata #<hash-table equal 0/65 0x2693c45> :watches #<hash-table equal 0/65 0x2693c65> :workspace-folders nil :last-id 0 :status-string nil :shutdown-action nil :diagnostics #<hash-table equal 0/65 0x2693c85>) . #<hash-table equal 1/65 0x2633719>)))
  #f(compiled-function () #<bytecode 0x2a23969>)()
  #f(compiled-function (result) #<bytecode 0x2a23d6d>)(#<hash-table equal 1/65 0x2633719>)
  lsp--parser-on-message(#s(lsp--parser :headers nil :body nil :reading-body nil :body-length nil :body-received nil :leftovers "" :workspace #s(lsp--workspace :parser #1 :server-capabilities nil :registered-server-capabilities nil :root "c:/Users/roey/src/topgrade/" :client #s(lsp--client :language-id nil :add-on? nil :new-connection (:connect #f(compiled-function (filter sentinel name) #<bytecode 0x294fcbd>) :test\? #f(compiled-function () #<bytecode 0x2951c51>)) :ignore-regexps nil :ignore-messages nil :notification-handlers #<hash-table equal 1/65 0x28a76dd> :request-handlers #<hash-table equal 0/65 0x2954cd1> :response-handlers #<hash-table eql 1/65 0x29ccf01> :prefix-function nil :uri-handlers #<hash-table equal 0/65 0x29ce351> :action-handlers #<hash-table equal 0/65 0x29cf01d> :use-native-json nil :major-modes (rust-mode rustic-mode) :activation-fn nil :priority -1 :server-id rls :multi-root nil :initialization-options nil :library-folders-fn nil :before-file-open-fn nil :initialized-fn nil :remote? nil :message-trace nil) :host-root nil :proc #<process rls> :cmd-proc #<process rls> :buffers (#<buffer generic.rs>) :highlight-overlays #<hash-table eq 0/65 0x2693b1d> :extra-client-capabilities nil :status starting :metadata #<hash-table equal 0/65 0x2693c45> :watches #<hash-table equal 0/65 0x2693c65> :workspace-folders nil :last-id 0 :status-string nil :shutdown-action nil :diagnostics #<hash-table equal 0/65 0x2693c85>)) "{\"jsonrpc\":\"2.0\",\"id\":1,\"result\":{\"capabilities\":{\"textDocumentSync\":2,\"hoverProvider\":true,\"completionProvider\":{\"resolveProvider\":true,\"triggerCharacters\":[\".\",\":\"]},\"definitionProvider\":true,\"implementationProvider\":true,\"referencesProvider\":true,\"documentHighlightProvider\":true,\"documentSymbolProvider\":true,\"workspaceSymbolProvider\":true,\"codeActionProvider\":true,\"codeLensProvider\":{\"resolveProvider\":false},\"documentFormattingProvider\":true,\"documentRangeFormattingProvider\":false,\"renameProvider\":true,\"executeCommandProvider\":{\"commands\":[\"rls.applySuggestion-6468\",\"rls.deglobImports-6468\"]}}}}")
  #f(compiled-function (proc output) #<bytecode 0x29da335>)(#<process rls> "Content-Length: 605\015\n\015\n{\"jsonrpc\":\"2.0\",\"id\":1,\"result\":{\"capabilities\":{\"textDocumentSync\":2,\"hoverProvider\":true,\"completionProvider\":{\"resolveProvider\":true,\"triggerCharacters\":[\".\",\":\"]},\"definitionProvider\":true,\"implementationProvider\":true,\"referencesProvider\":true,\"documentHighlightProvider\":true,\"documentSymbolProvider\":true,\"workspaceSymbolProvider\":true,\"codeActionProvider\":true,\"codeLensProvider\":{\"resolveProvider\":false},\"documentFormattingProvider\":true,\"documentRangeFormattingProvider\":false,\"renameProvider\":true,\"executeCommandProvider\":{\"commands\":[\"rls.applySuggestion-6468\",\"rls.deglobImports-6468\"]}}}}Content-Length: 92\015\n\015\n{\"jsonrpc\":\"2.0\",\"method\":\"window/progress\",\"params\":{\"id\":\"progress_1\",\"title\":\"Building\"}}Content-Length: 165\015\n\015\n{\"jsonrpc\":\"2.0\",\"method\":\"client/unregisterCapability\",\"id\":1,\"params\":{\"unregisterations\":[{\"id\":\"rls-range-formatting\",\"method\":\"textDocument/rangeFormatting\"}]}}Content-Length: 113\015\n\015\n{\"jsonrpc\":\"2.0\",\"method\":\"window/progress\",\"params\":{\"id\":\"progress_1\",\"message\":\"topgrade\",\"title\":\"Building\"}}Content-Length: 123\015\n\015\n{\"jsonrpc\":\"2.0\",\"method\":\"window/progress\",\"params\":{\"id\":\"progress_1\",\"message\":\"topgrade cfg(test)\",\"title\":\"Building\"}}")
```

**Comment by r-darwish** [comment]
If it helps, I changed `rest` to `cl-rest` in [this line](https://github.com/emacs-lsp/lsp-mode/commit/696295de214bdcb73fbabc9ed95095d6828cf7ad#diff-b1e9bc82c3b6bd6e95470e18ee41a19bR1608) and it seems to work.

**Comment by yyoncho** [comment]
It seems like 
```
(eval-when-compile
  (require 'cl))
```
does not work in some of the emacs versions. I changed the code to use cl-lib equivalent.

**Comment by r-darwish** [comment]
Thanks. Could [this line](https://github.com/emacs-lsp/lsp-mode/blob/f8c75dde93b77993e0a44ef6e7bb0e71dca4a160/lsp-mode.el#L1064) be a problem as well?

**Comment by yyoncho** [comment]
> Thanks. Could [this line](https://github.com/emacs-lsp/lsp-mode/blob/f8c75dde93b77993e0a44ef6e7bb0e71dca4a160/lsp-mode.el#L1064) be a problem as well?

Yes, I will change the 
```
(eval-when-compile
  (require 'cl))
```
to 
```
(require 'cl)
```

... untill I manage to do the proper fix since we have to fix first -> cl-first, second -> cl-second.

**Comment by yyoncho** [comment]
Using first/second is present in the code for a long time, I am not sure why the issue is reported now. - relevant commit in dap-mode https://github.com/yyoncho/dap-mode/commit/d7f79d019382936ef8618fd02bf5d5dbe169364f

**Comment by dakra** [comment]
@yyoncho I just want to make a minor nitpick. I update my packages from git and like to look through the changes, but when I real "Fix #1234" as the only commit message, that doesn't tell me anything. So it's better to describe why you changed something.  The "Fix #1234" remark can be in the end of the message body.. A good read is http://chris.beams.io/posts/git-commit/

Thanks for all you work
